Update Subnet Gateway Behavior

jtopjian opened the pull request:
This commit enables all three behaviors of a gateway during subnet creation and
updating.

If a GatewayIP is omitted, Neutron will provision a default gateway.
If a GatewayIP is set to an empty string, no gateway will be provisioned.
If a GatewayIP is specified, it will be used as the gateway IP.

A convenience function of `StringToPointer` was added to help pass empty
strings as pointers.

For #101

jtopjian commented:
@jrperritt ping for review/thoughts?

The [last time](https://github.com/rackspace/gophercloud/commit/f92ae6c762ef68e7fbe914f793d14356805a0a50) I came across this, I added the `NoGateway` field, so I'm trying something different this time.

I think this is just an area where any solution in Go might feel awkward. Some might think that omitting the GatewayIP means they don't want a gateway at all rather than to have Neutron make one.

If this isn't a good solution, maybe the `NoGateway` field should be revisited? `NoGateway` was emulating the `neutron` cli tool's `--no-gateway` flag, but maybe that's something too app-specific for Gophercloud?

jrperritt commented on params.go:
```diff
@@ -181,6 +181,12 @@ func IntToPointer(i int) *int {
 
+// StringToPointer is a function for converting strings into string pointers.
+// This is useful when passing in options to operations.
+func StringToPointer(s string) *string {
```
How is this different from `MaybeString` (the function below this one)?

jtopjian replied:
It's different because it's called `StringToPointer`. 

Yes, good catch. IIRC, I saw `MaybeString` and without reading it, disregarded it as something that would check if something was a string, bool, int, etc.

jtopjian replied:
Unless I'm mistaken, one difference is that `StringToPointer` will return `&""`, where `MaybeString` will not.

jrperritt replied:
> Unless I'm mistaken, one difference is that StringToPointer will return &"", where MaybeString will not.

Yes, I know :) I don't think we need this function. Creating a pointer to a `string` (as evidenced by this function) is trivial, and I suspect creating one for a `string` constant is uncommon. If you'd like something for the tests, you can define a exported variable in one of the common testing packages:
```go
var(
  Empty = ""
  EmptyPtr = &empty
)
```

jtopjian commented on openstack/networking/v2/subnets/requests.go:
```diff
@@ -79,7 +79,7 @@ type CreateOpts struct {
 	TenantID        string                `json:"tenant_id,omitempty"`
 	AllocationPools []AllocationPool      `json:"allocation_pools,omitempty"`
-	GatewayIP       *string               `json:"gateway_ip"`
+	GatewayIP       *string               `json:"-"`
```
The map should have three different forms:

1. User explicitly sets a gateway: 

`map[network_id:12345 cidr:1.2.3.0/24 gateway_ip:1.2.3.1]`

2. User does not want a gateway to exist: 

`map[network_id:12345 cidr:1.2.3.0/24 gateway_ip:]`

3. User wants Neutron to create a gateway, which will end up being `1.2.3.1`: 

`map[network_id:12345 cidr:1.2.3.0/24]`

jrperritt replied:
OK, adding `omitempty` to the struct tag:

```
requests_test.go:22: m: map[subnet:map[network_id:12345 cidr:1.2.3.4.]]
requests_test.go:32: m: map[subnet:map[network_id:12345 cidr:1.2.3.4. gateway_ip:]]
```

jtopjian replied:
I swear that was not working for me! I'll be cleaning up that other PR so I'll double-check this.

jtopjian replied:
It's not working for me with `omitempty`. 

To confirm: the only change you made to master was the addition of `omitempty` to the `GatewayIP` tag?

Are you printing the map after `omitempty` would have stripped away the nil value? Isn't that the point of `omitempty`?

I'm printing `b` right after it was created in `Create` and either it's omitted (because it was nil) or it has a value of `""` (which Neutron does not want).

jrperritt replied:
OK, I've done some testing, and I think I see where the discrepancy lies:

> 2) User does not want a gateway to exist:

> map[network_id:12345 cidr:1.2.3.0/24 gateway_ip:]

Not exactly. What we want is `gateway_ip:null` (or equivalently `gateway_ip:<nil>`). What's above represents `gateway_ip:""`, which OpenStack indeed does not accept.

With that in mind, I agree that this cannot be done as-is. Adding `omitempty` to the struct tag and modifying `ToSubnetCreateMap` to the following should fix it:

```go
func (opts CreateOpts) ToSubnetCreateMap() (map[string]interface{}, error) {
	b, err := gophercloud.BuildRequestBody(opts, "subnet")
	if err != nil {
		return nil, err
	}

	if m := b["subnet"].(map[string]interface{}); m["gateway_ip"] == "" {
		m["gateway_ip"] = nil
	}

	return b, nil
}
```

With those 2 changes:
- not providing a value for `CreateOpts.GatewayIP` will give the user the OpenStack default gateway
- providing the [a pointer to] the empty string will result in the subnet not having a gateway.

Now, is that reasonable or should it be flipped, where the user has to explicitly ask for the default gateway and otherwise gets no gateway?

jtopjian replied:
Cool - thank you for the attention on this. :)

I think the proposed workflow is good. This mirrors the default behavior in the Neutron CLI tool.

jtopjian commented:
@jrperritt `ToSubnetCreateMap` has been updated and `StringToPointer` has been removed.

jtopjian commented on openstack/networking/v2/subnets/requests.go:
```diff
@@ -112,15 +121,30 @@ type UpdateOptsBuilder interface {
 type UpdateOpts struct {
 	Name           string      `json:"name,omitempty"`
-	GatewayIP      string      `json:"gateway_ip,omitempty"`
+	GatewayIP      *string     `json:"-"`
```
Yes, totally forgot about Update. Coming up.

jtopjian commented:
@jrperritt Update updated, acceptance and unit tests added for gateway update.

jtopjian commented on openstack/networking/v2/subnets/requests.go:
```diff
@@ -112,15 +121,24 @@ type UpdateOptsBuilder interface {
 type UpdateOpts struct {
 	Name           string      `json:"name,omitempty"`
-	GatewayIP      string      `json:"gateway_ip,omitempty"`
+	GatewayIP      *string     `json:"gateway_ip,omitempty"`
```
In this case, I think reverting `UpdateOpts` is the best choice. While it makes for being inconsistent with `CreateOpts`, it more naturally enforces allowed values. I think that's the style more favored here.

I'll update the code, but won't push anything until confirmation.

jrperritt replied:
> So is the only option left simply to ignore the case in the code and document the behavior?

Yes, this `Update` should work like all the others.

jtopjian replied:
OK, so no changes needed to the Update part then?

If not, I have an extra unit test to push and then I think it's ready for another full review.

jrperritt replied:
I think what you have there is sufficient.

jrperritt commented on openstack/networking/v2/subnets/testing/requests_test.go:
```diff
@@ -276,6 +277,7 @@ func TestCreate(t *testing.T) {
 		NetworkID: "d32019d3-bc6e-4319-9c1d-6722fc136a22",
 		IPVersion: 4,
 		CIDR:      "192.168.199.0/24",
+		GatewayIP: gophercloud.MaybeString("192.168.199.1"),
```
`MaybeString` is intended for internal use (but created before an `internal` package was a thing). Because users refer to tests for direction, let's define an IP address as a constant (or variable, if it'll get changed) and take its address here.

jrperritt commented on openstack/networking/v2/subnets/testing/requests_test.go:
```diff
@@ -469,6 +556,64 @@ func TestUpdate(t *testing.T) {
+	var noGateway = ""
+	opts := subnets.UpdateOpts{
+		Name:      "my_new_subnet",
+		GatewayIP: &noGateway,
```
:+1:

jtopjian commented:
@jrperritt I've updated what should be the final commit (let me know if these should have been appended rather than squashed - I know you've asked for that before and I have probably slipped in that habit)

The two changes were: removal of `MaybeString` and I added an additional unit test I had not committed before.

jtopjian commented on acceptance/openstack/networking/v2/networking.go:
```diff
@@ -63,13 +63,12 @@ func CreateSubnet(t *testing.T, client *gophercloud.ServiceClient, networkID str
 		IPVersion:  4,
 		Name:       subnetName,
-		EnableDHCP: &iFalse,
+		EnableDHCP: gophercloud.Disabled,
```
@jrperritt Is this along the same lines of `MaybeString`? I found this to be a helpful both in terms of being a shortcut as well as readability ("DHCP is Disabled").

jrperritt commented:
> let me know if these should have been appended rather than squashed

A few things:
- firstly, it's OK; no changes needed here,
- secondly, the merger can (and should, for targeted PRs) now squash all commits in the PR upon merging,  
- thirdly, last item under `On Pull Requests` section of style guide: "Unless explicitly asked, do not squash commits in the middle of a review; only append. It makes it difficult for the reviewer to see what's changed from one review to the next."

jrperritt commented:
+2

jtopjian commented:
Thank you for your help on this one!

fatmcgav commented on openstack/networking/v2/subnets/requests.go:
```diff
@@ -88,7 +88,16 @@ type CreateOpts struct {
+	}
+
+	if m := b["subnet"].(map[string]interface{}); m["gateway_ip"] == "" {
+		m["gateway_ip"] = nil
```
@jtopjian / @jrperritt Shouldn't this line have been `b["gateway_ip"] = nil`?

Else how does `b["gateway_ip"]` get set to `nil`?
